crates/bevy_ptr/src/lib.rs

```diff
@@ -305,6 +322,7 @@ impl<'a, T> From<&'a [T]> for ThinSlicePtr<'a, T> {
 /// Creates a dangling pointer with specified alignment.
 /// See [`NonNull::dangling`].
 pub fn dangling_with_align(align: NonZeroUsize) -> NonNull<u8> {
+    assert!(align.is_power_of_two(), "Alignment must be power of two.");
```

Comment:
```suggestion
    debug_assert!(align.is_power_of_two(), "Alignment must be power of two.");
```

Reply:
This is a safety requirement. Unless we make this function unsafe, this should remain a non-debug assertion.

Reply:
It's only unsound to dereference an unaligned pointer, not to create one. It's perfectly acceptable to create a `NonNull<u8>` that is not a power of two.

Reply:
the intention behind  the `Ptr*` types is that they act the same as type erased borrows, so I think that we should be enforcing that alignment is correct on construction of `Ptr`/`PtrMut`/`OwningPtr` etc (via a `# Safety`). From this POV i think having it be a debug assertion (`fn debug_ensure_aligned` maybe to make it clear that it does not always assert) is fine

relatedly I don't think documenting these checks as `# Panic` is useful, they should be solely there to help catch UB that was reached on construction of the `Ptr` types.

Reply:
> the intention behind the Ptr* types is that they act the same as type erased borrows, so I think that we should be enforcing that alignment is correct on construction of Ptr/PtrMut/OwningPtr etc (via a # Safety). 

Being able to make unaligned pointers is useful though, see #7039 which makes use of `OwningPtr` to implement `CommandQueue` much more safely.

> relatedly I don't think documenting these checks as # Panic is useful

I agree.

Reply:
Removed the panic documentation.

>It's only unsound to dereference an unaligned pointer, not to create one. It's perfectly acceptable to create a NonNull<u8> that is not a power of two.

Alignment *must* be a power of two. It's unsafe to create a `Layout` that violates this. This is even checked with a [release mode panic in `ptr::is_aligned_to`](https://doc.rust-lang.org/src/core/ptr/const_ptr.rs.html#1358). If we're explicitly making a dangling pointer with a specified alignment, IMO we should always be ensuring that the alignment adheres to this.

Reply:
> Alignment _must_ be a power of two. [It's unsafe to create a `Layout` that violates this](https://doc.rust-lang.org/stable/std/alloc/struct.Layout.html#method.from_size_align_unchecked).

Right, but we're not creating a `Layout`, we're just creating a `NonNull<>`. Passing a non power of two to this function  is *incorrect*, but it's not unsound.

Let me put it like this: yes, alignment has to be a power of two, but `bevy_ptr::dangling_with_align` is the wrong place to check it. Adding a `debug_assert` would be helpful for catching errors early, but it's not required for safety reasons. Furthermore, simply making sure that it's a power of two doesn't really prove anything -- it's possible for an alignment to be a power of two and *still* be invalid, so we don't gain anything by stuffing runtime checks into this function.